File: frontend/src/interfaces/jobpost.ts
//หน้าที่: บอก React/TypeScript ว่า object ของโพสต์ มีโครงสร้างยังไง
import type { Employer } from "./employer";
import type { Student } from "./student";
import type { JobCategory } from "./job_category";
import type { EmploymentType } from "./employment_type";
import type { SalaryType } from "./salary_type";

export interface Jobpost {
  ID: number;
  CreatedAt: string;
  UpdatedAt: string;
  DeletedAt?: string | null;

  title: string;
  description: string;
  deadline: string;                
  status: string;
  image_url?: string | null;
  portfolio_required?: string | null;
  salary: number;
  locationjob: string;

  employer_id: number;
  Employer?: Employer;             

  job_category_id: number;
  JobCategory?: JobCategory;

  employment_type_id: number;
  EmploymentType?: EmploymentType;

  salary_type_id: number;
  SalaryType?: SalaryType;
}

// interfaces/jobpost.ts
export interface CreateJobpost {
  title: string;
  description: string;
  salary: number;
  locationjob: string;
  deadline: string;              
  status: string;                 
  portfolio_required?: string | null;     
  job_category_id: number;
  employment_type_id: number;
  salary_type_id: number;
  image_url?: string | null;
  employer_id: number;
}

// interface for myjob
export interface Jobpost {
    ID: number;
    CreatedAt: string;
    UpdatedAt: string;
    DeletedAt?: string | null;
    title: string;
    description: string;
    status: string;
    salary: number;
    employer_id: number;
    employer?: {
      company_name: string;
    };
    job_category_id: number;
    job_category?: JobCategory;
    locationjob: string;
    employment_type_id: number;
    employment_type?: EmploymentType;
    salary_type_id: number;
    salary_type?: SalaryType;
    student_id: number;
    student?: Student;
    payment_status_name?: string;
    ready_to_pay?: boolean;
}

File: backend/entity/jobpost.go
package entity

import (
	"time"

	"gorm.io/gorm"
)

type StatusEnum string

const (
	Open  StatusEnum = "Open"
	Close StatusEnum = "Close"
)

type Jobpost struct {
	gorm.Model
	Title             string     `gorm:"type:varchar(255);not null" json:"title"`
	Description       string     `gorm:"type:text;not null" json:"description"`
	Deadline          time.Time  `gorm:"not null" json:"deadline"`
	Status            StatusEnum `gorm:"not null" json:"status"`
	ImageURL          *string    `gorm:"type:varchar(255)" json:"image_url"`          // nullable string
	PortfolioRequired *string    `gorm:"type:varchar(100)" json:"portfolio_required"` // nullable
	Salary            int        `json:"salary"`
	LocationJob       string     `gorm:"column:locationjob;type:varchar(255);not null" json:"locationjob"`

	//FK
	EmployerID uint     `gorm:"not null" json:"employer_id"`
	Employer   Employer `gorm:"foreignKey:EmployerID;references:ID" json:"employer"`

	JobCategoryID uint        `json:"job_category_id"`
	JobCategory   JobCategory `gorm:"foreignKey:JobCategoryID"`

	// LocationID			uint			`gorm:"not null" json:"location_id"`
	// Location			Location		`gorm:"foreignKey:LocationID;references:ID"`

	EmploymentTypeID uint           `gorm:"not null" json:"employment_type_id"`
	EmploymentType   EmploymentType `gorm:"foreignKey:EmploymentTypeID;references:ID"`

	SalaryTypeID uint       `gorm:"not null" json:"salary_type_id"`
	SalaryType   SalaryType `gorm:"foreignKey:SalaryTypeID;references:ID"`

	//เหมือนจะไม่ได้ใช้แล้วเพราะว่ามันรู้ในหน้า jopappแล้วว่าสมัครโพสต์ไหน
	// StudentID			uint			`json:"student_id"`
	// Student				*Student		`gorm:"foreignKey: StudentID;references:ID" json:"student"`

	// เพิ่มตรงนี้ preload ผู้สมัครในโพสต์งาน
	Applications []JobApplication `gorm:"foreignKey:JobPostID" json:"applications,omitempty"`
}
